tests/integration/helpers.py

```diff
@@ -351,5 +355,46 @@ async def deploy_github_runner_charm(
         storage=storage,
     )
 
-    await model.wait_for_idle(status=ACTIVE, timeout=60 * 30)
+    if wait_idle:
+        await model.wait_for_idle(status=ACTIVE, timeout=60 * 30)
+
     return application
+
+
+async def wait_for(
+    func: Callable[[], Union[Awaitable, Any]],
+    timeout: int = 300,
+    check_interval: int = 10,
+) -> Any:
+    """Wait for function execution to become truthy.
+
+    Args:
+        func: A callback function to wait to return a truthy value.
+        timeout: Time in seconds to wait for function result to become truthy.
+        check_interval: Time in seconds to wait between ready checks.
+
+    Raises:
+        TimeoutError: if the callback function did not return a truthy value within timeout.
+
+    Returns:
+        The result of the function if any.
+    """
+    deadline = time.time() + timeout
+    is_awaitable = inspect.iscoroutinefunction(func)
+    while time.time() < deadline:
+        if is_awaitable:
+            if result := await func():
+                return result
+        else:
+            if result := func():
+                return result
+        time.sleep(check_interval)
+
+    # final check before raising TimeoutError.
+    if is_awaitable:
```

Comment:
Python doesn't quite have a do-while loop, the function above is of a do-while style, this just does a final check before raising an error just in case :D
Would I have understood you correctly here?

Reply:
You did, yes. Just ignore my comment